libs/sdk-core/src/breez_services.rs

```diff
@@ -773,31 +773,41 @@ impl BreezServices {
         invoice: Option<LNInvoice>,
         payment_res: Result<PaymentResponse>,
     ) -> SdkResult<Payment> {
-        if payment_res.is_err() {
-            self.notify_event_listeners(BreezEvent::PaymentFailed {
-                details: PaymentFailedData {
-                    error: payment_res.as_ref().err().unwrap().to_string(),
-                    node_id,
-                    invoice,
-                },
-            })
-            .await?;
-            return Err(SdkError::SendPaymentFailed {
-                err: payment_res.err().unwrap().to_string(),
-            });
-        }
-        let payment = payment_res.unwrap();
-        self.do_sync(true).await?;
-
-        match self.persister.get_payment_by_hash(&payment.payment_hash)? {
-            Some(p) => {
-                self.notify_event_listeners(BreezEvent::PaymentSucceed { details: p.clone() })
-                    .await?;
-                Ok(p)
+        match payment_res {
+            Ok(payment) => {
+                self.do_sync(true).await?;
+
+                match self.persister.get_payment_by_hash(&payment.payment_hash)? {
+                    Some(p) => {
+                        self.notify_event_listeners(BreezEvent::PaymentSucceed {
+                            details: p.clone(),
+                        })
+                        .await?;
+                        Ok(p)
+                    }
+                    None => Err(SdkError::SendPaymentFailed {
+                        err: "Payment not found".into(),
+                    }),
+                }
+            }
+            Err(e) => {
+                if let Some(inv) = invoice.clone() {
```

Comment:
I think the approach to persist failed payment should be different. In multi device use case it will end up persisting only in the sender device, the other devices won't have this data.
If we sync failed payments from the remote node and add this logic as part of the sync (pull_changes) it will ensure they will end up on all devices. I suggest to take a look here: https://github.com/breez/breez-sdk/blob/main/libs/sdk-core/src/greenlight/node_api.rs#L867

Reply:
I've adapted `pull_transactions` to use the CLN client and pb's to list invoice and payments. This exposes the erroronion in case of a failed payment. It seems like the erroronion is always none though for any failed payments I have.